srtcore/sync.cpp

```diff
@@ -98,6 +98,115 @@ int64_t get_cpu_frequency()
 
 const int64_t s_cpu_frequency = get_cpu_frequency();
 
+CSync::CSync(pthread_cond_t& cond, CGuard& g)
+    : m_cond(&cond), m_mutex(g.mutex())
+{
+    // XXX it would be nice to check whether the owner is also current thread
+    // but this can't be done portable way.
+
+    // When constructed by this constructor, the user is expected
+    // to only call signal_locked() function. You should pass the same guard
+    // variable that you have used for construction as its argument.
+}
+
+CSync::CSync(pthread_cond_t& cond, Mutex& mutex, Nolock)
+    : m_cond(&cond)
+    , m_mutex(&mutex)
+{
+    // We expect that the mutex is NOT locked at this moment by the current thread,
+    // but it is perfectly ok, if the mutex is locked by another thread. We'll just wait.
+
+    // When constructed by this constructor, the user is expected
+    // to only call lock_signal() function.
+}
+
+void CSync::wait()
+{
+    THREAD_PAUSED();
+    pthread_cond_wait(&(*m_cond), &m_mutex->ref());
+    THREAD_RESUMED();
+}
+
+bool CSync::wait_until(const steady_clock::time_point& exptime)
+{
+    // This will work regardless as to which clock is in use. The time
+    // should be specified as steady_clock::time_point, so there's no
+    // question of the timer base.
+    steady_clock::time_point now = steady_clock::now();
+    if (now >= exptime)
+        return false; // timeout
+
+    THREAD_PAUSED();
+    bool signaled = SyncEvent::wait_for(m_cond, &m_mutex->ref(), exptime - now) != ETIMEDOUT;
+    THREAD_RESUMED();
+
+    return signaled;
+}
+
+/// Block the call until either @a timestamp time achieved
+/// or the conditional is signaled.
+/// @param [in] delay Maximum time to wait since the moment of the call
+/// @retval true Resumed due to getting a CV signal
+/// @retval false Resumed due to being past @a timestamp
+bool CSync::wait_for(const steady_clock::duration& delay)
+{
+    // Note: this is implemented this way because the pthread API
+    // does not provide a possibility to wait relative time. When
+    // you implement it for different API that does provide relative
+    /// time waiting, you may want to implement it better way.
+
+    THREAD_PAUSED();
+    bool signaled = SyncEvent::wait_for(m_cond, &m_mutex->ref(), delay) != ETIMEDOUT;
+    THREAD_RESUMED();
+
+    return signaled;
+}
+
+void CSync::lock_signal()
+{
+    // We expect m_nolock == true.
+    lock_signal(*m_cond, *m_mutex);
```

Comment:
`m_nolock` is something that will be added in the future? Or in the debug logic?

Reply:
Right. The comment remained after I purged this from the thread logging implementation stuff.